File: engine-scoring/src/main/kotlin/com/palmastro/scoring/ScoringEngineImpl.kt
package com.palmastro.scoring

import com.palmastro.contracts.*
import com.palmastro.contracts.interfaces.ScoringEngine

class ScoringEngineImpl(
    private val ruleset: Ruleset = Ruleset.default()
) : ScoringEngine {

    private val domains = listOf("career", "wealth", "family", "health")
    private val baseline = 50

    override fun score(input: ScoringInput): ScoringResult {
        val palmSignals = SignalResolver.resolvePalmSignals(input.palmFeatures, ruleset)
        val astroSignals = SignalResolver.resolveAstroSignals(input.astroResult, ruleset)

        val confMultiplier = ruleset.confidenceMultipliers[input.palmFeatures.confidence] ?: 0.5

        val domainScores = mutableMapOf<String, Int>()
        val explainability = mutableListOf<ExplainEntry>()

        for (domain in domains) {
            var score = baseline.toDouble()

            for (signal in palmSignals) {
                if (meetsConfidence(signal.minConfidence, input.palmFeatures.confidence)) {
                    val weight = signal.domainWeights[domain] ?: 0.0
                    val contribution = signal.direction * signal.magnitude * weight * confMultiplier
                    score += contribution
                    if (weight > 0.3) {
                        explainability.add(ExplainEntry(signal.signalId, "${signal.signalId} → $domain", contribution))
                    }
                }
            }

            for ((signal, astro) in astroSignals) {
                val astroConf = ruleset.confidenceMultipliers[astro.confidence] ?: 0.5
                val weight = signal.domainWeights[domain] ?: 0.0
                val contribution = signal.direction * signal.magnitude * weight * astroConf
                score += contribution
                if (weight > 0.3) {
                    explainability.add(ExplainEntry(signal.signalId, "${signal.signalId} → $domain", contribution))
                }
            }

            domainScores[domain] = score.toInt().coerceIn(0, 100)
        }

        val overallScore = domainScores.values.average().toInt()
        val grade = assignGrade(overallScore)
        val confidence = minConfidence(input.palmFeatures.confidence, astroConfidence(input.astroResult))
        val reasons = buildConfidenceReasons(input)

        val sortedExplain = explainability
            .distinctBy { "${it.signalId}-${it.mapping}" }
            .sortedByDescending { kotlin.math.abs(it.contribution) }

        return ScoringResult(
            domainScores = domainScores,
            subdimScores = emptyMap(),
            grade = grade,
            confidence = confidence,
            confidenceReasons = reasons,
            explainability = sortedExplain,
            matchedBuckets = emptyList(),
            rulesetVersion = ruleset.version
        )
    }

    private fun assignGrade(score: Int): String {
        for ((grade, range) in ruleset.gradeIntRanges) {
            if (score in range) return grade
        }
        return ruleset.gradeIntRanges.keys.last()
    }

    private fun meetsConfidence(required: String, actual: String): Boolean {
        val order = mapOf("low" to 0, "med" to 1, "high" to 2)
        return (order[actual] ?: 0) >= (order[required] ?: 0)
    }

    private fun astroConfidence(astro: AstroResult): String =
        if (astro.calcLevel == CalcLevel.L2) "high" else "med"

    private fun minConfidence(a: String, b: String): String {
        val order = mapOf("low" to 0, "med" to 1, "high" to 2)
        return if ((order[a] ?: 0) <= (order[b] ?: 0)) a else b
    }

    private fun buildConfidenceReasons(input: ScoringInput): List<String> {
        val reasons = mutableListOf<String>()
        if (input.palmFeatures.confidence == "low") reasons.add("scan_quality_low")
        if (input.palmFeatures.featureCoverage < 0.5f) reasons.add("low_feature_coverage")
        if (input.astroResult.calcLevel == CalcLevel.L1) reasons.add("missing_birth_time")
        return reasons
    }
}

File: ios/PalmAstroKit/Sources/ScoringEngine/ScoringEngineImpl.swift
import Foundation
import CoreContracts

/// Port of engine-scoring ScoringEngineImpl.kt — the arithmetic must stay
/// bit-for-bit equivalent so scores match across platforms (PRD §10, §18).
public final class ScoringEngineImpl: ScoringEngineProtocol {

    private let ruleset: Ruleset
    private let domains = Domains.all
    private let baseline = 50

    public init(ruleset: Ruleset) {
        self.ruleset = ruleset
    }

    /// Convenience initializer loading the bundled default ruleset.
    public convenience init() throws {
        self.init(ruleset: try Ruleset.loadDefault())
    }

    public func score(input: ScoringInput) -> ScoringResult {
        let palmSignals = SignalResolver.resolvePalmSignals(features: input.palmFeatures, ruleset: ruleset)
        let astroSignals = SignalResolver.resolveAstroSignals(astro: input.astroResult, ruleset: ruleset)

        let confMultiplier = ruleset.confidenceMultipliers[input.palmFeatures.confidence] ?? 0.5

        var domainScores: [String: Int] = [:]
        var explainability: [ExplainEntry] = []

        for domain in domains {
            var score = Double(baseline)

            for signal in palmSignals where meetsConfidence(required: signal.minConfidence, actual: input.palmFeatures.confidence) {
                let weight = signal.domainWeights[domain] ?? 0.0
                let contribution = Double(signal.direction) * Double(signal.magnitude) * weight * confMultiplier
                score += contribution
                if weight > 0.3 {
                    explainability.append(ExplainEntry(
                        signalId: signal.signalId,
                        mapping: "\(signal.signalId) → \(domain)",
                        contribution: contribution
                    ))
                }
            }

            for (signal, astro) in astroSignals {
                let astroConf = ruleset.confidenceMultipliers[astro.confidence] ?? 0.5
                let weight = signal.domainWeights[domain] ?? 0.0
                let contribution = Double(signal.direction) * Double(signal.magnitude) * weight * astroConf
                score += contribution
                if weight > 0.3 {
                    explainability.append(ExplainEntry(
                        signalId: signal.signalId,
                        mapping: "\(signal.signalId) → \(domain)",
                        contribution: contribution
                    ))
                }
            }

            domainScores[domain] = min(max(Int(score), 0), 100)
        }

        let overallScore = domainScores.isEmpty
            ? baseline
            : Int(Double(domainScores.values.reduce(0, +)) / Double(domainScores.count))
        let grade = ruleset.grade(forScore: overallScore)
        let confidence = minConfidence(input.palmFeatures.confidence, astroConfidence(input.astroResult))
        let reasons = buildConfidenceReasons(input: input)

        // Kotlin's sortedByDescending is stable; Swift's sort is not guaranteed
        // to be, so break abs-contribution ties by insertion order explicitly —
        // parity fixtures compare explainability entry order exactly.
        let sortedExplain = distinctByKey(explainability) { "\($0.signalId)-\($0.mapping)" }
            .enumerated()
            .sorted { a, b in
                let absA = abs(a.element.contribution)
                let absB = abs(b.element.contribution)
                return absA != absB ? absA > absB : a.offset < b.offset
            }
            .map(\.element)

        return ScoringResult(
            domainScores: domainScores,
            subdimScores: [:],
            grade: grade,
            confidence: confidence,
            confidenceReasons: reasons,
            explainability: sortedExplain,
            matchedBuckets: [],
            rulesetVersion: ruleset.version
        )
    }

    // MARK: - Helpers (mirroring the Kotlin implementation)

    private func meetsConfidence(required: String, actual: String) -> Bool {
        let order = ["low": 0, "med": 1, "high": 2]
        return (order[actual] ?? 0) >= (order[required] ?? 0)
    }

    private func astroConfidence(_ astro: AstroResult) -> String {
        astro.calcLevel == .L2 ? "high" : "med"
    }

    private func minConfidence(_ a: String, _ b: String) -> String {
        let order = ["low": 0, "med": 1, "high": 2]
        return (order[a] ?? 0) <= (order[b] ?? 0) ? a : b
    }

    private func buildConfidenceReasons(input: ScoringInput) -> [String] {
        var reasons: [String] = []
        if input.palmFeatures.confidence == "low" { reasons.append("scan_quality_low") }
        if input.palmFeatures.featureCoverage < 0.5 { reasons.append("low_feature_coverage") }
        if input.astroResult.calcLevel == .L1 { reasons.append("missing_birth_time") }
        return reasons
    }

    private func distinctByKey<T>(_ items: [T], key: (T) -> String) -> [T] {
        var seen = Set<String>()
        var result: [T] = []
        for item in items {
            let k = key(item)
            if seen.insert(k).inserted {
                result.append(item)
            }
        }
        return result
    }
}
